### ISSUE TITLE
DBS APIs that sum over the number of events

### ISSUE THREAD

**Issue by yuyiguo** [dmwm/DBS#541]
@ericvaandering @ticoann @hufnagel @vkuznet 

DBS API, such as listFileSummaries return the total number of events for a given block or dataset, when the client provides the optional run information, the API will pick only the runs in the given block or dataset. Currently we only store event count by files, if there are more than one runs per file, this API will only give an estimate event numbers.

If we store the event counts per lumi, then we will be able to give the exact event counts in a given run for a give block or dataset. The problem is when a dataset has mix data that is some blocks have event per lumi, but others only have event per file. The API has to examine the event count lumi by lumi in order to give the exact number.  The newly added checking on lumi by lumi will slow down the API a lot.  Because we have 300 second limit on cmsweb, we need to put some thought on this.

If we have to query on the event_count before we can decide what to do, we need to add index on event count and this index will slow down the insert.

**Comment by ericvaandering** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Define DBS behavior RW with and without provided data](https://trello.com/c/m81LEKep/16-define-dbs-behavior-rw-with-and-without-provided-data)

**Comment by ericvaandering** [comment]
Can you paste the exact query that caused the "problem" here? And indicate if the run number in that query can also be a list?

**Comment by ericvaandering** [comment]
IIRC @hufnagel also had some trick for an index column that is null/not null that might be useful

**Comment by hufnagel** [comment]
Thing is, sum of event per run is imprecise now. With mixed records in the future it'll stay imprecise, so there isn't any requirement to get it correct either. It'll only be correct if we 100% detailed information.

And yes, would be good to see the query.

**Comment by ericvaandering** [comment]
Sort of. It's correct 99% of the time now.…

**Comment by yuyiguo** [comment]
I don't have the queries yet because I don't know what we want to achieve with the mixed data. what is the accept imprecise.

**Comment by hufnagel** [comment]
You have the current queries. We aren't going to write new queries to first order, but we might want to tune existing queries to give better results. So what are the existing queries that would be effected by this ? Anything that deals with nevents in a run or lumi specific way would fall into this category.

**Comment by yuyiguo** [comment]
https://github.com/dmwm/DBS/blob/master/Server/Python/src/dbs/dao/Oracle/File/SummaryList.py#L26

**Comment by yuyiguo** [comment]
@ericvaandering @hufnagel 
This is the new query that intents to count the number of events in a given  block/dataset for  a give runs when we add event_count per lumi.  There are several requirements in order to give the correct number:
1. The same run, lumi pair can only appears once in the queried block/dataset.
2. In single file, either all the run, lumi pairs have event count per lumi or all NULL.
3. This relaxed the requirement from all have event count per lumi or all have NULL in a block to a file.

The query is only for show the idea, will need to rearrange if this is what we intent to do.

select sum(et) from (
with setcs as (select b.block_id, FL1.file_id, run_num, sum(FL1.event_count) sect
FROM FILE_LUMIS FL1, blocks b, files f
where FL1.file_id=f.file_id and f.block_id=b.block_id and b.block_id=1 
and run_num in(43, 9999999)
group by b.block_id, FL1.file_id, run_num 
order by fl1.file_id asc)
select  setcs.block_id, setcs.file_id, run_num,
(case 
when setcs.sect is NULL then
(select f.event_count from files f where setcs.file_id=f.file_id) 
else setcs.sect
end) et from setcs
);

**Comment by yuyiguo** [comment]
improved one:
select sum(sumet) num_event from
(
with setcs as (select FL.file_id, run_num, sum(FL.event_count) et
FROM FILE_LUMIS FL
join files f on f.file_id=fl.file_id
join blocks b on b.block_id=f.block_id
where b.block_id=1 and run_num in(43, 9999999)
group by fl.file_id, fl.run_num 
)
select 
(case 
when setcs.et is NULL then
(select f.event_count from files f where f.file_id=setcs.file_id) 
else setcs.et
end) sumet from setcs
);

**Comment by ericvaandering** [comment]
You can't assume #1, unfortunately.

**Comment by ericvaandering** [comment]
#2 is safe though

**Comment by yuyiguo** [comment]
if not #1, then we need to think about how accurate the total event count will be.  
Say R1, L1 has 100 events in f1 and R1, L1 has 200 events in f2.  How should we count the total events in f1+f2?

**Comment by yuyiguo** [comment]
This one takes care one file has more than one runs, but no events per lumi is filled.

select sum(sumet) num_event from
(
with setcs as (select FL.file_id,  sum(FL.event_count) et
FROM FILE_LUMIS FL
join files f on f.file_id=fl.file_id
join blocks b on b.block_id=f.block_id
where b.block_id=1 and run_num in(43,45, 9999999)
group by fl.file_id
)
select 
(case 
when setcs.et is NULL then
(select f.event_count from files f where f.file_id=setcs.file_id) 
else setcs.et
end) sumet from setcs
);

**Comment by ericvaandering** [comment]
It's 300 in that case. That's kind of what I was arguing for somehow indicating that listFileSummaries is not for counting events at the sub-file level at all and that to get the proper count, we need to supply a dedicated API that really counts events in runs or run/lumi combinations with a dataset or block as a restrictor. 

Of course we can describe this in the docs. Figuring out how to do it in the API short of dropping the info is not reasonable.

**Comment by ericvaandering** [comment]
In other words listFileSummaries is by design supposed to answer the question: "How many events are in files that contain run=6 in dataset=xyz"

Another API is is needed to answer "How many events are in run=6 in dataset=xyz" or "How many events are in run=6, lumi=5 in dataset=xyz"

The fact that most of the time the answer to the first question is the same as the answer to the second question is a happy accident.

**Comment by yuyiguo** [comment]
@ericvaandering @hufnagel 
1. Keep listFileSummaries as it is. It is the file level summary. It answers total number of events (and others) in the files belong to a dataset, if the optional run=[1,2,3,...] is used, then it will only count the events belong to these runs. If R1, L1 appears in more than one files, it will be count more than ones(e1 (for the events in R1,L1 in file 1)+e2(for the events in R1,L1 in file 2)). If one file has more than one runs, this API cannot distinguish  how to divide the events between these two runs by definition. This API cannot be use to query number of events by individual runs, then add them together to be the total numbers.

2. If needed, we can add a new API, say listFileLumiEventSummaries. This new API intents to count total events at lumi, run level. We need to define:
-  The scope of the API. I would say per block.
- the behavior of the API when the lumi level events is not available.

**Comment by ericvaandering** [comment]
What do you mean with the first bullet point (scope of the API)? Do you mean it would not be possible for the user to query for a whole dataset at a time but would have to go block by block? If so, will there be a way in the client to do this automatically for the user?

**Comment by vkuznet** [comment]
Eric,
certain API prohibit to use dataset as an input and require either block or
lfn. In DAS I hide this from users, i.e. user provide query with dataset=/a/b/c
and under the hood I resolve dataset name into set of blocks/lfns and
then send N-requests to DBS to get the info.

I think Yuyi meant similar thing, we'll need to resolve dataset into
blocks in a client and query block by block. Client can do that
sequentially or concurrently.

**Comment by hufnagel** [comment]
listFileSummaries gives you information for files only. You can restrict to files belonging to certain runs, but you still get file level information, not run level information, hence you also get files where only part of the file belongs to a certain run. That's all fine, I don't think anyone wants to change anything.

At the same time, this gives file level information and as such, it should return the full run/lumi/event map for the file, shouldn't it ? Like it returns the run/lumi content now we just expand that.

Given that, DAS has all the knowledge about what's in the file and any kind of aggregation and remapping of that information into a run or lumi context can happen within DAS.

What I don't see here is where there is a need for a new DBS API. If DAS already has all the information, why do we need another DBS API ? Do we foresee a use case where someone asks for run/lumi specific aggregate numbers where we would not expand this and pull fully detailed dataset/block/file info from DBS ?

**Comment by ericvaandering** [comment]
I think @hufnagel is right. While we've been encouraging people to use the DBS client for queries that only require DBS and not other services, all the info should be there. The user can sum if they want.

One question, though? What happens with the current listFileSummaries API when run # is specified. Does it still show that a file has information from a run that was /not/ specified if the file has two runs? It seems to me that that should be the correct behavior.

**Comment by yuyiguo** [comment]
@vkuznet  @ericvaandering @hufnagel

The new API intends give lumi level data so it could be very large. I don't think one need this data for entire dataset for most of time. In order to protect DBS and for the API to stay in the CMSWEB timeout limit, I would do it only at a block . If a dataset is needed in a few cases, adding the blocks together is not hard.   
A lot of scripts use DBS directly so we cannot depend on DAS to do all the aggregation.
listFileSummaries is not for the run/lumi/event mapping. We have been trying very hard not to put a lot of information into one DBS API. DBS2 approved that would not scale. ListFileLumi API gives the mapping.
The current listFileSummaries API when run # is specified will add all the events in the files even if a file has more than one runs. In other words, the API take all events in a file as whole even if there may be only part of the file should be counted.

**Comment by hufnagel** [comment]
run/lumi/event mapping is part of the file metadata. As such, retrieving this in the context of file information is correct IMO. You could argue that this should not be the default level of detail provided by listFileSummaries, but there should be an option to get it. We don't have another API to read file metadat, don't we ? So listFileSummaries has to provide at least an option to select everything we know about a file.

**Comment by hufnagel** [comment]
IMO providing access to run/lumi/event only via specialized API calls that already do the aggregation is the wrong approach. We simple don't know how users will use this information and as such, we need to provide access to the raw information and let them aggregate themself.

**Comment by hufnagel** [comment]
And a that point you have to ask yourself, given that we have a low level API to provide the raw data, how many specialized API calls do we really need ? What are the use cases for them ?

**Comment by hufnagel** [comment]
@ericvaandering , answering your question. IMO the run parameter specified in the listFileSummaries call is a filter. Give me all files that contain data for this run. No guarantee is made that these files will *only* contain that run.

**Comment by vkuznet** [comment]
Dirk,
here I'll be very careful. Even though I agree with you comment, I'll give you
example of poor data management in CMS which already may impact on request
you're asking for.

There are huge datasets in DBS with large number of files, the largest I've
seen contains 125K files. Let me assume that user want to know detailed
information about such large dataset. How do we envision to do it efficiently
on a client side? It has 25K blocks, and if client want to know aggregated
information about entire dataset it should place 25K queries to DBS in order to
get aggregated information (should it be done in DAS or elsewhere it does not
matter in this context). This is an example, but I think you get my point. I
rather prefer aggregation happen on DB side since client request can be very
expensive to support.

**Comment by yuyiguo** [comment]
I have a question.  In the context of event_count per lumi, what is difference between the event_count=0 and is NULL?  In the current schema, event_count per file is not NULL.

For the new data, we will calculate the event_count per file from event_count per lumi. If the event_count per lumi is NULL as we discussed, what would be the event_count per file?

**Comment by yuyiguo** [comment]
@vkuznet @hufnagel 
Data in a file  and in lumi are two very different levels in DBS. There is no one API can get all of them. The listFileSummaries API is the file level API so one may use it with a dataset. However when one asks run/lumi/event mapping, the lumi level data, if the dataset is huge, such as in Valentin's example, 25k blocks/125k files, how many run/lumi/event map will be?  This kind API will no scale. It will time out or slow down everything else.

**Comment by hufnagel** [comment]
@vkuznet we'll never be able to completely rule out users asking for stupid things.

The huge problem with aggregation on the server side is that to be able to do it, you need to *completely* understand the use cases, ie. in what context users will want to access and aggregate this information. Which IMHO at the moment we have little to no clue about. As such, I think there simple is no way not provide a method for users to access the raw information.

Also, the current API already returns the run/lumi map for a file, correct ? Adding an event count to that doesn't add that much information. As such, any unreasonable query of such type already is unreasonable now.

**Comment by hufnagel** [comment]
@yuyiguo run/lumi/event with NULL events means we don't know how many events, we just know that the run/lumi is in that file. OTOH, 0 events means that the run/lumi is present in that file with 0 events. Very different things.

EDM data isn't just event data. The file contains a run/lumi but the event selection efficiency was such that 0 events were selected is completely different from the file not containing the run/lumi at all.

**Comment by yuyiguo** [comment]
@hufnagel The point is we cannot have an API to give run/lumi/event map for a dataset. But it is ok to have it for a file or block.  This is already done.

**Comment by hufnagel** [comment]
@yuyiguo on your second comment, no one is advocating that we add queries to retrieve run/lumi information. Our primary access method remains file based. As far as I understand our discussions so far, we will never ask a question that starts with run/lumi.

You will always start with "given this dataset, block or otherwise selected set of files, what can you tell me about their content related to a given run or lumi.

Which means in terms of selection in the database, you always select your files first and then you look at the file/run/lumi map and apply whatever aggregation or filters you want.

All we add here is more detailed information by adding another level to the map. File/run/lumi becomes file/run/lumi/event, which means we allow different aggregation and sums and such.

The question then is, do we have a system where we select the files, pull all the raw information for the files and let whatever system pulls that info do the aggregation (DAS or a user script) or do we have short-cut methods within DBS that skip pulling all the detailed file level information and aggregate directly server-side.

IMHO we simply don't understand enough about all the possible use cases for now for the latter option to be a valid and workable approach.

**Comment by yuyiguo** [comment]
@hufnagel if I have map: file=20, run=10, lumi=1, event=0 in file_lumis table, do these mean I have run/lumi in this file, but event count is 0? How this is different from file=20, run=10, lumi=1, event is NULL?

**Comment by hufnagel** [comment]
@yuyiguo event=0 means the process that produced that file was processing data for run=10, lumi=1 ut selected 0 events. In a skim for instance this could esily happen. If event=NULL it means we don't know how many events were selected, we could have 0, 10, 100, 1000, whatever events.

**Comment by hufnagel** [comment]
For the physics side, knowing that the file was produced from run=10, lumi=1 is essential, even if no events are actually in the file. They need to know selection efficiency.

**Comment by yuyiguo** [comment]
@hufnagel OK, but how could we didn't know how many events were selected after the processing? 
Then what does the event_count=0 mean in the file table? The file table is not NULL for the event_count per file.

**Comment by hufnagel** [comment]
event_count=0 means the file contains run=10,lumi=1 with 0 events. Having 0 events for a run/lumi in an EDM file is valid. It's completely different from the file file not containing the run/lumi.

Think of EDM files as containing events and run/lumis and these two are somewhat independent.

**Comment by vkuznet** [comment]
> Also, the current API already returns the run/lumi map for a file, correct ? Adding an event count to that doesn't add that much information. As such, any unreasonable query of such type already is unreasonable now.

Dirk, let's clear this up with concrete examples. Yuyi may provide server side
API names, here I list REST APIs accessible via URL:

DBS API filesummaries API returns:

```
https://cmsweb.cern.ch/dbs/prod/global/DBSReader/filesummaries?block_name=/JetHT/CMSSW_7_4_12_patch1-2015_09_15_HLTnewconditions0_74X_dataRun2_HLTValidation_forPostTS2_v1-v1/RAW%23a20b5f06-5c0e-11e5-90ef-02163e00d7b3

[{"num_file": 1, "num_lumi": 13, "num_block": 1, "num_event": 5512, "file_size": 1075948154}]
```

while filelumis API returns:

```
https://cmsweb.cern.ch/dbs/prod/global/DBSReader/filelumis?block_name=/JetHT/CMSSW_7_4_12_patch1-2015_09_15_HLTnewconditions0_74X_dataRun2_HLTValidation_forPostTS2_v1-v1/RAW%23a20b5f06-5c0e-11e5-90ef-02163e00d7b3


[{"lumi_section_num": [80, 85, 79, 81, 86, 93, 82, 88, 90, 78, 87, 83, 84], "logical_file_name": "/store/relval/CMSSW_7_4_12_patch1/JetHT/RAW/2015_09_15_HLTnewconditions0_74X_dataRun2_HLTValidation_forPostTS2_v1-v1/00000/FE8224E6-0D5C-E511-BC4F-0026189438D9.root", "run_num": 254790}]
```

Since you want run/lumi/event_count raw output it means that later API should split
output for EVERY lumi number, in every run and return something like that:

```
[
{"lumi_section_num":80, "event_count":1, "logical_file_name":"lfn", "run_num": 254790},
{"lumi_section_num":85, "event_count":2, "logical_file_name":"lfn", "run_num": 254790},
{"lumi_section_num":79, "event_count":3, "logical_file_name":"lfn", "run_num": 254790},
...
]
```

Instead of having single row you'll get in this case 13 (equal to # lumis)
rows.  If dataset is large and we need to process large number of blocks, take
this number as N, your output (or output client should handle) would be:

N x number_of_lumis_in_given x number_of_runs_in_given_file

It would be nice to get a sense how large such output would be and in fact,
verify that data response time will be manageable.
Remember, DBS server will need to construct this output on a fly
and since DBS deals with dicts (which directly translate into RAM usage)
I don't know how well DBS API will handle such requests.

**Comment by hufnagel** [comment]
@vkuznet I don't quite follow your example. Why do we have to stick with a flat dict structure ? Is that a limitation in the return format of the DBS REST API for some reason ?

Something like this would save a ton of memory

{ 'lfn': 'blah', 'run': ((10,1,100),(10,2,100),(10,3,100),(11,1,100)) }

for a file containing run=0 and lumis 1-3 and run=11 and lumi 1, with every lumi containing 100 events.  Isn't something like this (or a multi-dimensional dict structure if the set isn't doable) possible ?

**Comment by ericvaandering** [comment]
Yeah

```
[{"lumi_section_num": [80, 85, 79, 81, 86, 93, 82, 88, 90, 78, 87, 83, 84], "logical_file_name": "/store/relval/CMSSW_7_4_12_patch1/JetHT/RAW/2015_09_15_HLTnewconditions0_74X_dataRun2_HLTValidation_forPostTS2_v1-v1/00000/FE8224E6-0D5C-E511-BC4F-0026189438D9.root", "run_num": 254790}]
```
 is already not complete for a file since a file can have more than one run (the thing that precipitated this already. But if nesting structures is not possible and the size is really a concern, then simply adding 
```
"events_per_lumi": [100,101, ..., ]``` to the above is a possibility (although you then have to be careful to keep both lists in the same order.

**Comment by vkuznet** [comment]
> @vkuznet I don't quite follow your example. Why do we have to stick with a flat dict structure ? Is that a limitation in the return format of the DBS REST API for some reason ?

Yuyi may correct me, but I think the flat format comes from direct mapping of
ORACLE look-up values to dict attributes, i.e.

run,lfn,lumi
1,bla,1

is mapped as {"run":1, "lfn":"bla", "lumi":1}

DBS seems to do some optimization to combine lumis into single list.


> 
> Something like this would save a ton of memory
> 
> { 'lfn': 'blah', 'run': ((10,1,100),(10,2,100),(10,3,100),(11,1,100)) }

this is fine and should work, but it is change of API return output
and will require some work on DBS server side to collect, group run/lumi/events
triplets out of ORACLE dump.

> for a file containing run=0 and lumis 1-3 and run=11 and lumi 1, with every lumi containing 100 events.  Isn't something like this (or a multi-dimensional dict structure if the set isn't doable) possible ?


and, remember, nested dict structures are really bad performant in python
since python allocates for every nest dict a memory chunk.

Let Yuyi clarify details:
- why flat structure is used
- how much effort will be to re-structure ORACLE output to compact one keeping
  in mind large number of output rows

**Comment by hufnagel** [comment]
I have written REST APIs that rely on Oracle queries. Sure, what you get back from the query is fully expanded and flat. There is no requirement to return it exactly like that though, you can always remap into whatever data structure you want before returning.

This is python code, restructuring in memory should be fast. Plus you gain savings in not having to transfer the huge flat structure overhead through the REST API layers.

But as you said, changing the return format of a REST APU call is tricky since you'll break all old clients. But then, we are planning to change the return format somewhat anyways, so that part is maybe not the deal breaker it would be otherwise.

**Comment by hufnagel** [comment]
And there are migration strategies for such changes as well. We could add an optional flag to the REST API call that requests the new format and then any new client uses that and the old clients get the old format.

**Comment by yuyiguo** [comment]
I have to make slides for tomorrow's meeting. So just quick answers to your questions. I hope you all can go to the meeting tomorrow so we can discuss it there.

As I pointed out earlier, listFileSummaries is a file level summary API that gives summary info as Valentin pointed in his example. The individual detailed lumi/run/event data does not fit there.

DBS return a list of dictionaries as the direct Oracle query result set in most of time, but in some cases, we do combined the result, such as  listfileLumis API returns:  [{"lumi_section_num": [80, 85, 79], "logical_file_name": "0026189438D9.root", "run_num": 254790}, {"lumi_section_num": [80, 85, 79], "logical_file_name": "0026189438D9.root", "run_num": 254791}].

If flat structure or combined one depends on what is the output. If there is not much repeated info, DBS will keep it flat to speed up. If there is a lot of repeated data, then DBS will combined. In addition, DBS API is used for migration server too, If the we nested the output, we will have to expand them before to insert into another database. So it is a balance.

**Comment by ericvaandering** [comment]
@yuyiguo So you are already breaking the return format on file, run pairs effectively. That's in the REST. What does the DBS python API do with this? Does it just return it that way as well or does it reassemble it into a structure where LFN is a property of the file and lumis is a property of the run?

**Comment by yuyiguo** [comment]
DBS Client APIs return the same as the server does.

**Comment by yuyiguo** [comment]
Done.
